src/pudl/io_managers.py

```diff
@@ -0,0 +1,592 @@
+        """
+        table_name = self._get_table_name(context)
+        _ = self._get_sqlalchemy_table(table_name)
+
+        engine = self.engine
+
+        with engine.connect() as con:
+            return pudl.metadata.fields.apply_pudl_dtypes(
+                pd.read_sql_table(table_name, con)
+            )
+
+
+@io_manager(
+    config_schema={
+        "pudl_output_path": Field(
+            EnvVar(
+                env_var="PUDL_OUTPUT",
+            ),
+            description="Path of directory to store the database in.",
+            default_value=None,
+        ),
+    }
+)
+def pudl_sqlite_io_manager(init_context) -> SQLiteIOManager:
+    """Create a SQLiteManager dagster resource for the pudl database."""
+    base_dir = init_context.resource_config["pudl_output_path"]
+    md = Package.from_resource_ids().to_sql()
+    return SQLiteIOManager(base_dir=base_dir, db_name="pudl", md=md)
+
+
+class FercSQLiteIOManager(SQLiteIOManager):
+    """IO Manager for reading tables from FERC databases.
+
+    This class should be subclassed and the load_input and handle_output
+    methods should be implemented.
+
+    This IOManager exepcts the database to already exist.
+    """
+
+    def __init__(
+        self, base_dir: str = None, db_name: str = None, md: sa.MetaData = None
+    ):
+        """Initialize FercSQLiteIOManager."""
+        super().__init__(base_dir, db_name, md)
+
+    def _setup_database(self) -> sa.engine.Engine:
+        """Create database engine and read the metadata.
+
+        Returns:
+            engine: SQL Alchemy engine that connects to a database in the base_dir.
+        """
+        # If the sqlite directory doesn't exist, create it.
+        db_path = self.base_dir / f"{self.db_name}.sqlite"
+        if not db_path.exists():
+            raise ValueError(
+                f"No {self.db_name}.sqlite found. Run the job that creates the {self.db_name} database."
+            )
+
+        engine = sa.create_engine(f"sqlite:///{db_path}")
+
+        # Connect to the local SQLite DB and read its structure.
+        ferc1_meta = sa.MetaData()
+        ferc1_meta.reflect(engine)
+        self.md = ferc1_meta
+
+        return engine
+
+    def handle_output(self, context: OutputContext, obj):
+        """Handle an op or asset output."""
+        raise NotImplementedError(
+            "FercSQLiteIOManager can't write outputs. Subclass FercSQLiteIOManager and implement the handle_output method."
+        )
+
+    def load_input(self, context: InputContext) -> pd.DataFrame:
+        """Load a dataframe from a sqlite database.
+
+        Args:
+            context: dagster keyword that provides access output information like asset name.
+        """
+        raise NotImplementedError(
+            "FercSQLiteIOManager can't load inputs. Subclass FercSQLiteIOManager and implement the load_input method."
+        )
+
+
+class FercDBFSQLiteIOManager(FercSQLiteIOManager):
```

Comment:
clarification: we take the DBF and the XBRL data from FERC and load them into SQLite with two different schemas, and these two IOManagers grab data from the two different schemas?

Reply:
I created separate IOManagers for the XBRL and DBF data because the FERC ETL extracts the data from these two databases using different SQL statements. For XBRL, the some fields from the `identification_001_duration` table need to be merged in. See the `load_input` methods for the two classes.